### ISSUE TITLE
Existing variations broken after 2.3.11 -> 2.4.6 in a test environment.

### ISSUE THREAD

**Issue by janiszalitis** [woothemes/woocommerce#8966]
Hi!

I have a "clean" environment for testing Woocommerce data import.
It has Wordpress, default, 2015 theme, Wocoomerce and WP All import plugin with Woocommerce addon (for import testing).
0 other plugins, no specific configuration changes, no custom code except for a data processing hook for the import plugin.

I also have a development website with custom theme and a bunch of plugins.
After upgrading 2.3.11->2.4.6 today I noticed that, for variable products, the admin interface went from showing 1 attribute for each variation (normal, as configured by me before), to that 1 attribute + ALL the other attributes prefixed by "Any", like "Any size", "Any color", etc.

After scratching my head for a while, I decided to upgrade the test environment in the same way.
This way I am sure it's an issue with 2.4.6, because I did no other changes to test environment and it suddenly showed the same issue.

How it looked in 2.3.11:
![screen shot 2015-08-27 at 14 44 44](https://cloud.githubusercontent.com/assets/1185430/9520387/7bb34128-4ccf-11e5-893d-34b56d49aa6d.jpg)

How the same product looks now in 2.4.6:
![screen shot 2015-08-27 at 14 44 20](https://cloud.githubusercontent.com/assets/1185430/9520396/87e7a8f8-4ccf-11e5-862b-31c503dad59b.jpg)

Since it's a site in development in final stages, it is critical issue for me and I started checking out what is going on.
What I found thus far:
1. Data are 100% correct and unchanged. Also, they were identical between the broken and unbroken versions before I upgraded the unbroken one to broken. Attribute value "is_variation" is set up correctly, i.e. only one attribute has it true (the one visible in first screenshot).
2. The AJAX that calls the display for this screen also correctly puts "is_variation=false" for all those "surplus" attributes which are now showing on variation tab (and shouldn't).

I tried to find where in woocoommerce this output is calculated, to pin down the exact line of issue, but have not yet.

Any help appreciated.
Thank you!

**Comment by sandissauka** [comment]
Thanks @janiszalitis for posting this. I have a similar issue after upgrading to 2.4.6. Looking forward to hear from Woo people.

**Comment by janiszalitis** [comment]
I just noticed I've uploaded screenshots from different installations. Here's a screenshot from the same system as #1 screenshot in my issue, but after upgrade.

![screen shot 2015-08-27 at 15 30 56](https://cloud.githubusercontent.com/assets/1185430/9520522/b92c55d4-4cd0-11e5-99d5-4540c08b2845.jpg)

**Comment by aaronbwcollins** [comment]
I also had to manually fix variations and composite products. After each upgrade I have to remove then add each product.

**Comment by mikejolley** [comment]
@janiszalitis you imported this data, are you certain is_variation is setup as it should be?

Can you see if this commit resolves https://github.com/woothemes/woocommerce/commit/1f3e8b984e84e3c97d3490ed8c6db5b2bf170d5e

**Comment by janiszalitis** [comment]
1. No, this doesn't change anything.

2. If I add print_r($attributes) before line 606, the output comes out when I open the product, but not when I open variation tab. So it may be that there is misunderstanding between us. The problem appears when open variation tab.

Print_r before line 606 gives me this:
<pre>Array
(
    [pa_izmers] =&gt; Array
        (
            [name] =&gt; pa_izmers
            [value] =&gt; 
            [position] =&gt; 0
            [is_visible] =&gt; 1
            [is_variation] =&gt; 1
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_razotajs] =&gt; Array
        (
            [name] =&gt; pa_razotajs
            [value] =&gt; 
            [position] =&gt; 0
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_ramis-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_ramis-velosipeda
            [value] =&gt; 
            [position] =&gt; 0
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_kam-paredzets] =&gt; Array
        (
            [name] =&gt; pa_kam-paredzets
            [value] =&gt; 
            [position] =&gt; 1
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_aizmugurejais-parsledzejs] =&gt; Array
        (
            [name] =&gt; pa_aizmugurejais-parsledzejs
            [value] =&gt; 
            [position] =&gt; 1
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_parsledzeju-rokturis] =&gt; Array
        (
            [name] =&gt; pa_parsledzeju-rokturis
            [value] =&gt; 
            [position] =&gt; 2
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_sezona] =&gt; Array
        (
            [name] =&gt; pa_sezona
            [value] =&gt; 
            [position] =&gt; 2
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_bremzes-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_bremzes-velosipeda
            [value] =&gt; 
            [position] =&gt; 3
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_centra-ass-klani-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_centra-ass-klani-velosipeda
            [value] =&gt; 
            [position] =&gt; 4
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_aploces-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_aploces-velosipeda
            [value] =&gt; 
            [position] =&gt; 5
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_rumbas-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_rumbas-velosipeda
            [value] =&gt; 
            [position] =&gt; 6
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_riepas-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_riepas-velosipeda
            [value] =&gt; 
            [position] =&gt; 7
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

    [pa_svars-velosipeda] =&gt; Array
        (
            [name] =&gt; pa_svars-velosipeda
            [value] =&gt; 
            [position] =&gt; 8
            [is_visible] =&gt; 1
            [is_variation] =&gt; 
            [is_taxonomy] =&gt; 1
            [is_create_taxonomy_terms] =&gt; 1
        )

)
</pre>

Here are the POST params from AJAX call.
<pre>action=woocommerce_load_variations&security=cb47a00497&product_id=16464&attributes%5Bpa_izmers%5D%5Bname
%5D=pa_izmers&attributes%5Bpa_izmers%5D%5Bvalue%5D=&attributes%5Bpa_izmers%5D%5Bposition%5D=0&attributes
%5Bpa_izmers%5D%5Bis_visible%5D=1&attributes%5Bpa_izmers%5D%5Bis_variation%5D=1&attributes%5Bpa_izmers
%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_izmers%5D%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_razotajs
%5D%5Bname%5D=pa_razotajs&attributes%5Bpa_razotajs%5D%5Bvalue%5D=&attributes%5Bpa_razotajs%5D%5Bposition
%5D=0&attributes%5Bpa_razotajs%5D%5Bis_visible%5D=true&attributes%5Bpa_razotajs%5D%5Bis_variation%5D
=false&attributes%5Bpa_razotajs%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_razotajs%5D%5Bis_create_taxonomy_terms
%5D=1&attributes%5Bpa_ramis-velosipeda%5D%5Bname%5D=pa_ramis-velosipeda&attributes%5Bpa_ramis-velosipeda
%5D%5Bvalue%5D=&attributes%5Bpa_ramis-velosipeda%5D%5Bposition%5D=0&attributes%5Bpa_ramis-velosipeda
%5D%5Bis_visible%5D=true&attributes%5Bpa_ramis-velosipeda%5D%5Bis_variation%5D=false&attributes%5Bpa_ramis-velosipeda
%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_ramis-velosipeda%5D%5Bis_create_taxonomy_terms%5D=1&attributes
%5Bpa_kam-paredzets%5D%5Bname%5D=pa_kam-paredzets&attributes%5Bpa_kam-paredzets%5D%5Bvalue%5D=&attributes
%5Bpa_kam-paredzets%5D%5Bposition%5D=1&attributes%5Bpa_kam-paredzets%5D%5Bis_visible%5D=true&attributes
%5Bpa_kam-paredzets%5D%5Bis_variation%5D=false&attributes%5Bpa_kam-paredzets%5D%5Bis_taxonomy%5D=1&attributes
%5Bpa_kam-paredzets%5D%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_aizmugurejais-parsledzejs%5D%5Bname
%5D=pa_aizmugurejais-parsledzejs&attributes%5Bpa_aizmugurejais-parsledzejs%5D%5Bvalue%5D=&attributes
%5Bpa_aizmugurejais-parsledzejs%5D%5Bposition%5D=1&attributes%5Bpa_aizmugurejais-parsledzejs%5D%5Bis_visible
%5D=true&attributes%5Bpa_aizmugurejais-parsledzejs%5D%5Bis_variation%5D=false&attributes%5Bpa_aizmugurejais-parsledzejs
%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_aizmugurejais-parsledzejs%5D%5Bis_create_taxonomy_terms%5D=1&attributes
%5Bpa_parsledzeju-rokturis%5D%5Bname%5D=pa_parsledzeju-rokturis&attributes%5Bpa_parsledzeju-rokturis
%5D%5Bvalue%5D=&attributes%5Bpa_parsledzeju-rokturis%5D%5Bposition%5D=2&attributes%5Bpa_parsledzeju-rokturis
%5D%5Bis_visible%5D=true&attributes%5Bpa_parsledzeju-rokturis%5D%5Bis_variation%5D=false&attributes%5Bpa_parsledzeju-rokturis
%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_parsledzeju-rokturis%5D%5Bis_create_taxonomy_terms%5D=1&attributes
%5Bpa_sezona%5D%5Bname%5D=pa_sezona&attributes%5Bpa_sezona%5D%5Bvalue%5D=&attributes%5Bpa_sezona%5D%5Bposition
%5D=2&attributes%5Bpa_sezona%5D%5Bis_visible%5D=true&attributes%5Bpa_sezona%5D%5Bis_variation%5D=false
&attributes%5Bpa_sezona%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_sezona%5D%5Bis_create_taxonomy_terms%5D
=1&attributes%5Bpa_bremzes-velosipeda%5D%5Bname%5D=pa_bremzes-velosipeda&attributes%5Bpa_bremzes-velosipeda
%5D%5Bvalue%5D=&attributes%5Bpa_bremzes-velosipeda%5D%5Bposition%5D=3&attributes%5Bpa_bremzes-velosipeda
%5D%5Bis_visible%5D=true&attributes%5Bpa_bremzes-velosipeda%5D%5Bis_variation%5D=false&attributes%5Bpa_bremzes-velosipeda
%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_bremzes-velosipeda%5D%5Bis_create_taxonomy_terms%5D=1&attributes
%5Bpa_centra-ass-klani-velosipeda%5D%5Bname%5D=pa_centra-ass-klani-velosipeda&attributes%5Bpa_centra-ass-klani-velosipeda
%5D%5Bvalue%5D=&attributes%5Bpa_centra-ass-klani-velosipeda%5D%5Bposition%5D=4&attributes%5Bpa_centra-ass-klani-velosipeda
%5D%5Bis_visible%5D=true&attributes%5Bpa_centra-ass-klani-velosipeda%5D%5Bis_variation%5D=false&attributes
%5Bpa_centra-ass-klani-velosipeda%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_centra-ass-klani-velosipeda%5D
%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_aploces-velosipeda%5D%5Bname%5D=pa_aploces-velosipeda
&attributes%5Bpa_aploces-velosipeda%5D%5Bvalue%5D=&attributes%5Bpa_aploces-velosipeda%5D%5Bposition%5D
=5&attributes%5Bpa_aploces-velosipeda%5D%5Bis_visible%5D=true&attributes%5Bpa_aploces-velosipeda%5D%5Bis_variation
%5D=false&attributes%5Bpa_aploces-velosipeda%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_aploces-velosipeda
%5D%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_rumbas-velosipeda%5D%5Bname%5D=pa_rumbas-velosipeda
&attributes%5Bpa_rumbas-velosipeda%5D%5Bvalue%5D=&attributes%5Bpa_rumbas-velosipeda%5D%5Bposition%5D
=6&attributes%5Bpa_rumbas-velosipeda%5D%5Bis_visible%5D=true&attributes%5Bpa_rumbas-velosipeda%5D%5Bis_variation
%5D=false&attributes%5Bpa_rumbas-velosipeda%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_rumbas-velosipeda%5D
%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_riepas-velosipeda%5D%5Bname%5D=pa_riepas-velosipeda
&attributes%5Bpa_riepas-velosipeda%5D%5Bvalue%5D=&attributes%5Bpa_riepas-velosipeda%5D%5Bposition%5D
=7&attributes%5Bpa_riepas-velosipeda%5D%5Bis_visible%5D=true&attributes%5Bpa_riepas-velosipeda%5D%5Bis_variation
%5D=false&attributes%5Bpa_riepas-velosipeda%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_riepas-velosipeda%5D
%5Bis_create_taxonomy_terms%5D=1&attributes%5Bpa_svars-velosipeda%5D%5Bname%5D=pa_svars-velosipeda&attributes
%5Bpa_svars-velosipeda%5D%5Bvalue%5D=&attributes%5Bpa_svars-velosipeda%5D%5Bposition%5D=8&attributes
%5Bpa_svars-velosipeda%5D%5Bis_visible%5D=true&attributes%5Bpa_svars-velosipeda%5D%5Bis_variation%5D
=false&attributes%5Bpa_svars-velosipeda%5D%5Bis_taxonomy%5D=1&attributes%5Bpa_svars-velosipeda%5D%5Bis_create_taxonomy_terms
%5D=1&page=1&per_page=10</pre>

**Comment by janiszalitis** [comment]
Here's the _product_attributes meta for this product.

<pre>a:13:{s:9:"pa_izmers";a:7:{s:4:"name";s:9:"pa_izmers";s:5:"value";s:0:"";s:8:"position";i:0;s:10:"is_visible";i:1;s:12:"is_variation";i:1;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:11:"pa_razotajs";a:7:{s:4:"name";s:11:"pa_razotajs";s:5:"value";s:0:"";s:8:"position";i:0;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:19:"pa_ramis-velosipeda";a:7:{s:4:"name";s:19:"pa_ramis-velosipeda";s:5:"value";s:0:"";s:8:"position";i:0;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:16:"pa_kam-paredzets";a:7:{s:4:"name";s:16:"pa_kam-paredzets";s:5:"value";s:0:"";s:8:"position";i:1;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:28:"pa_aizmugurejais-parsledzejs";a:7:{s:4:"name";s:28:"pa_aizmugurejais-parsledzejs";s:5:"value";s:0:"";s:8:"position";i:1;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:23:"pa_parsledzeju-rokturis";a:7:{s:4:"name";s:23:"pa_parsledzeju-rokturis";s:5:"value";s:0:"";s:8:"position";i:2;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:9:"pa_sezona";a:7:{s:4:"name";s:9:"pa_sezona";s:5:"value";s:0:"";s:8:"position";i:2;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:21:"pa_bremzes-velosipeda";a:7:{s:4:"name";s:21:"pa_bremzes-velosipeda";s:5:"value";s:0:"";s:8:"position";i:3;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:30:"pa_centra-ass-klani-velosipeda";a:7:{s:4:"name";s:30:"pa_centra-ass-klani-velosipeda";s:5:"value";s:0:"";s:8:"position";i:4;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:21:"pa_aploces-velosipeda";a:7:{s:4:"name";s:21:"pa_aploces-velosipeda";s:5:"value";s:0:"";s:8:"position";i:5;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:20:"pa_rumbas-velosipeda";a:7:{s:4:"name";s:20:"pa_rumbas-velosipeda";s:5:"value";s:0:"";s:8:"position";i:6;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:20:"pa_riepas-velosipeda";a:7:{s:4:"name";s:20:"pa_riepas-velosipeda";s:5:"value";s:0:"";s:8:"position";i:7;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}s:19:"pa_svars-velosipeda";a:7:{s:4:"name";s:19:"pa_svars-velosipeda";s:5:"value";s:0:"";s:8:"position";i:8;s:10:"is_visible";b:1;s:12:"is_variation";b:0;s:11:"is_taxonomy";i:1;s:24:"is_create_taxonomy_terms";i:1;}}</pre>

**Comment by mikejolley** [comment]
I don't know if you have a plugin affecting this, but I've confirmed there is no issue in master regarding this. 

The attributes are passed to our scripts via ajax (all attributes).

This code is then ran which outputs each select box *only if the attribute is for a variation*.

https://github.com/woothemes/woocommerce/blob/master/includes/admin/meta-boxes/views/html-variation-admin.php#L25-L26

This will fail if empty or 0 - it only outputs when is_variation is 1.

You could do a var_dump at that point and see what you have?

**Comment by janiszalitis** [comment]
Ok, we have found the problem.
Turns out, the "is_variation" flag is "false", which is a string and therefore checking for it with an "!" operator does not work.
The "true" value for the same flag is listed as "1", which reacts to "!" operator correctly, the end result being that the "!" operator thinks that both "1" and "false" are TRUE, inverts those to FALSE and the check in line 25 is simply skipped for all attributes regardless whether they are variations or not.

This case is described here, see intersection of last row and last (rightmost) column in the comparison table:
http://php.net/manual/en/types.comparisons.php

Here's the code as changed by me:
<pre>
			echo '<pre>';
			print_r($parent_data['attributes']);
			echo '</pre>';
		
			foreach ( $parent_data['attributes'] as $attribute ) {
				
				echo (int)(! $attribute['is_variation']) . '<br />';
				
				// Only deal with attributes that are variations
				if ( ! $attribute['is_variation'] ) {
					continue;
				}</pre>

And here's 2 screenshots what that gives.
1. The Print_r results: https://dl.dropboxusercontent.com/u/5288379/Woocommerce/Print_r.jpg
2. The "watch this!" results: https://dl.dropboxusercontent.com/u/5288379/Woocommerce/Result.jpg

Here's my proposed fix (line 25):
<pre>if ( ! $attribute['is_variation'] ||  $attribute['is_variation'] == 'false') {</pre>

Fix further explained here: http://stackoverflow.com/questions/7336861/how-to-convert-string-to-boolean-php
